### PR TITLE
Add jax_triton implementation of normalization op

### PR DESCRIPTION
**Pull request by samanklesaria**
This PR ports the pallas_triton implementation normalization to raw triton, as pallas_triton will be deprecated. Currently, the added test only compares the jax_triton implementation to the pallas_triton one, and does not port replicate the full set of pallas_triton tests for jax_triton.

### CODE REVIEW

**Comment by samanklesaria** [comment]
# Normalization: matching pallas_triton with the jax_triton 3D kernel

Benchmarks on NVIDIA A100-SXM4-40GB, JAX 0.11.0, jax-triton 0.3.1. Times are
CUPTI median device time in microseconds, heuristic configs (no autotuning).
Shapes are the AlphaFold workloads from `arg_specs.py`; "default axis" means the
normalization axis is the last one (axis=-1).

## Benchmarks


| Shape | forward: jt / pallas / xla | fwd+VJP: jt / pallas / xla |
|---|---|---|
| 384res_64chan       | **28** / 30 / 31   | **88** / 85 / 187 |
| 384res_128chan      | **58** / 60 / 60   | **176** / 174 / 367 |
| 768res_128chan      | **221** / 222 / 223 | **616** / 614 / 1347 |
| 384res_128chan_axis0| **62** / 67 / 144  | **190** / 185 / 408 |
| 768res_128chan_axis0| **232** / 239 / 495 | **655** / 670 / 1474 |

## Reference: shapes

| Name | x_shape | axis | dtype | canonical (M, A, N) |
|---|---|---|---|---|
| 384res_64chan | (384, 384, 64) | -1 | bf16 | (147456, 64, 1) |
| 384res_128chan | (384, 384, 128) | -1 | bf16 | (147456, 128, 1) |
| 768res_128chan | (768, 768, 128) | -1 | bf16 | (589824, 128, 1) |
| 384res_128chan_axis0 | (128, 384, 384) | 0 | bf16 | (1, 128, 147456) |
| 768res_128chan_axis0 | (128, 768, 768) | 0 | bf16 | (1, 128, 589824) |

File: tokamax/_src/ops/normalization/jax_triton.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+# ── Forward kernel ─────────────────────────────────────────────────────────────
+
+
+@triton.jit
+def _normalization_kernel(
+    # --- pointer args: inputs then outputs ---
+    X, SCALE, OFFSET,
+    Y, MEAN, RSTD,
+    # --- scalar args ---
+    M_dim, A, N_dim, stride_m, stride_a, epsilon, scale_offset,
```
Comment (superbobry) [inline tokamax/_src/ops/normalization/jax_triton.py]: I think all of these also need to be constexprs?

File: tokamax/_src/ops/normalization/jax_triton.py [hunk 2]
```diff
@@ -0,0 +1,497 @@
+# Copyright 2025 DeepMind Technologies Limited. All Rights Reserved.
```
Comment (superbobry) [inline tokamax/_src/ops/normalization/jax_triton.py]: 2026

File: tokamax/_src/ops/normalization/jax_triton.py [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+# ── VJP kernel ──────────────────────────────────────────────────────────────────
+
+
+@triton.jit
+def _normalization_vjp_kernel(
+    # --- pointer args: inputs then outputs ---
+    DOUT, X, SCALE, MEAN, RSTD,
+    DX, DSCALE, DOFFSET,
+    # --- scalar args ---
+    M_dim, A, N_dim, stride_m, stride_a, scale_offset, grid_n,
```
Comment (superbobry) [inline tokamax/_src/ops/normalization/jax_triton.py]: As above, I think these have to be constexprs?

File: tokamax/_src/ops/normalization/jax_triton_test.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
+# Copyright 2025 DeepMind Technologies Limited. All Rights Reserved.
```
Comment (superbobry) [inline tokamax/_src/ops/normalization/jax_triton_test.py]: 2026

File: tokamax/_src/ops/normalization/jax_triton_test.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  def __init__(self, *args):
+    super().__init__(*args, norm_fn=jax_triton.JaxTritonNormalization())
+
+  def setUp(self):
+    if jax.default_backend() == 'tpu':
+      self.skipTest('Not supported on TPUs.')
+    super().setUp()
+
+
+class JaxTritonVsPallasTritonTest(parameterized.TestCase):
```
Comment (superbobry) [inline tokamax/_src/ops/normalization/jax_triton_test.py]: I think we should fork existing tests for the Pallas Triton kernel instead, since the plan is to get rid of `PallasTritonNormalization` soon.

File: tokamax/_src/ops/normalization/jax_triton.py (hunk 1 above)
Comment (samanklesaria) [reply tokamax/_src/ops/normalization/jax_triton.py]: Certainly agree for `A`, especially as we're passing in the constexpr `BLOCK_A` derived from it. For `epsilon` and `scale_offset`, what would baking the float into the add at compile time vs using the runtime value gain us? Wouldn't it be the same cost either way? For the rest (batch and stride dimensions), is the idea that by baking in these values we'd get better vectorization? Could you help me understand the advantage?

**Comment by samanklesaria** [comment]
## Batching (`vmap`): fold-into-M vs. `lax.map`

`jt.triton_call` has no `jax.vmap` batching rule (it raises by design — the rule
is inherently per-kernel), so `JaxTritonNormalization` supplies one via
`jax.custom_batching.custom_vmap`. Two implementations were compared:

- **`lax.map`**: map the single-instance launch over the
  batch axis — B sequential kernel launches.
- **fold-into-M**: merge the batch into the leading row dimension and issue one
  batch-aware launch (grid axis 0 = `B * cdiv(M, BLOCK_M)`, tiling each batch
  element separately; `scale`/`offset` indexed per batch). Gated by a `BATCHED`
  constexpr so the non-`vmap` path stays codegen-identical.

Batched benchmarks (`*_vmap` in `bench.py`), batch size **B=8**, same A100 /
median device µs. jax_triton fold vs. lax.map, with pallas as reference:

**Forward**

| Shape (vmap, B=8) | fold µs | lax.map µs | fold speedup | pallas µs |
|---|---|---|---|---|
| 384res_64chan        | 224  | 330  | 1.47× | 221 |
| 384res_128chan       | 444  | 778  | 1.75× | 438 |
| 768res_128chan       | 1748 | 2755 | 1.58× | 1733 |
| 384res_128chan_axis0 | 493  | 788  | 1.60× | 473 |
| 768res_128chan_axis0 | 1954 | 3018 | 1.54× | 1872 |

**Forward + VJP**

| Shape (vmap, B=8) | fold µs | lax.map µs | fold speedup | pallas µs |
|---|---|---|---|---|
| 384res_64chan        | 622  | 873  | 1.40× | 622 |
| 384res_128chan       | 1214 | 1835 | 1.51× | 1205 |
| 768res_128chan       | 4698 | 6575 | 1.40× | 4679 |
| 384res_128chan_axis0 | 1449 | 1775 | 1.22× | 1300 |
| 768res_128chan_axis0 | 5337 | 6708 | 1.26× | 5134 |

**Takeaways:**

- Fold-into-M is **~1.4–1.75× faster** than `lax.map` on the batched path (one
  launch over `B*M` rows vs. B sequential launches).
- Fold-into-M brings jax_triton **on par with pallas** under `vmap` (within ~5%
  on forward; VJP trails pallas by ~1.1–1.4× on the axis0 shapes). `lax.map` was
  1.5–1.7× *slower* than pallas.
- **Non-`vmap` path is unchanged**: e.g. `384res_64chan` forward is 27.3 µs
  (fold) vs. 26.9 µs (lax.map) — within noise — confirming the `BATCHED=False`
  path is codegen-identical.
- Tradeoff of fold-into-M: it materializes `scale`/`offset` to `(B, A)` under
  `vmap` even when shared, and adds a batch offset to the block-pointer base
  (only on the batched path).

**Comment by samanklesaria** [comment]
## Constexpr dims/strides: no device-time benefit

Would baking the shape/stride/grid params (`M_dim`, `N_dim`, `stride_m`,
`stride_a`, `grid_n`) into the kernel as `tl.constexpr` — instead of passing them
as runtime scalars — generate faster code? Empirically, **no**. Same A100, same
batched benchmarks, jax_triton medians µs, runtime scalars vs. constexpr:

**Forward**

| Shape | runtime µs | constexpr µs |
|---|---|---|
| 384res_64chan             | 27.3 | 26.8 |
| 384res_128chan            | 57.7 | 57.4 |
| 768res_128chan            | 221  | 221  |
| 384res_128chan_axis0      | 63.8 | 63.6 |
| 768res_128chan_axis0      | 234  | 233  |
| 384res_64chan_vmap        | 224  | 225  |
| 384res_128chan_vmap       | 444  | 444  |
| 768res_128chan_vmap       | 1748 | 1756 |
| 384res_128chan_axis0_vmap | 493  | 490  |
| 768res_128chan_axis0_vmap | 1954 | 1958 |

**Forward + VJP**

| Shape | runtime µs | constexpr µs |
|---|---|---|
| 384res_64chan             | 86.6 | 87.2 |
| 384res_128chan            | 178  | 182  |
| 768res_128chan            | 623  | 623  |
| 384res_128chan_axis0      | 187  | 188  |
| 768res_128chan_axis0      | 650  | 649  |
| 384res_64chan_vmap        | 622  | 623  |
| 384res_128chan_vmap       | 1214 | 1216 |
| 768res_128chan_vmap       | 4698 | 4701 |
| 384res_128chan_axis0_vmap | 1449 | 1453 |
| 768res_128chan_axis0_vmap | 5337 | 5321 |

Every pair is within run-to-run noise (<2%, σ/median ≈ 0.01). Triton's built-in
divisibility specialization already secures the alignment/vectorization proofs,
so constexpr dims/strides add nothing
